app/main.py

```diff
@@ -1,34 +1,91 @@
+        return decks
 
-    def fire(self, row, column):
-        # Change the `is_alive` status of the deck
-        # And update the `is_drowned` value if it's needed
-        pass
+    def get_deck(self, row: int, column: int) -> Deck | None:
+        for deck in self.decks:
+            if deck.row == row and deck.column == column:
+                return deck
+        return None
+
+    def fire(self, row: int, column: int) -> str:
+        deck = self.get_deck(row, column)
+        if deck and deck.is_alive:
+            deck.is_alive = False
+            self.is_drowned = all(not d.is_alive for d in self.decks)
+            return "Sunk!" if self.is_drowned else "Hit!"
+        return "Miss!"
 
 
 class Battleship:
-    def __init__(self, ships):
-        # Create a dict `self.field`.
-        # Its keys are tuples - the coordinates of the non-empty cells,
-        # A value for each cell is a reference to the ship
-        # which is located in it
-        pass
-
-    def fire(self, location: tuple):
-        # This function should check whether the location
-        # is a key in the `self.field`
-        # If it is, then it should check if this cell is the last alive
-        # in the ship or not.
-        pass
+    def __init__(self, ships: list) -> None:
+        self.field = {}
+        self.ships = []
+        self._validate_field(ships)  # Validate the ship placements
+        for start, end in ships:
+            ship = Ship(start, end)
+            self.ships.append(ship)
+            for deck in ship.decks:
+                self.field[(deck.row, deck.column)] = ship
+
+    @staticmethod
+    def _validate_field(ships: list) -> None:
+        field = [[0] * 10 for _ in range(10)]
+
+        # Place ships and check overlap
+        for start, end in ships:
+            if start[0] == end[0]:  # Horizontal ship
+                row = start[0]
+                for col in range(start[1], end[1] + 1):
+                    if field[row][col] != 0:
+                        raise ValueError("Ships cannot overlap or touch.")
+                    field[row][col] = 1
+            elif start[1] == end[1]:  # Vertical ship
+                col = start[1]
+                for row in range(start[0], end[0] + 1):
+                    if field[row][col] != 0:
+                        raise ValueError("Ships cannot overlap or touch.")
+                    field[row][col] = 1
+
+        # Check neighboring cells
+        for row in range(10):
+            for col in range(10):
+                if field[row][col] == 1:
+                    for dr in (-1, 0, 1):
+                        for dc in (-1, 0, 1):
+                            nr, nc = row + dr, col + dc
+                            if (0 <= nr < 10 and 0 <= nc < 10
+                                    and field[nr][nc] == 1):
+                                if (dr, dc) != (0, 0) and field[nr][nc] != 2:
+                                    field[nr][nc] = 2  # Mark as neighboring
```

Comment:
The logic for checking neighboring cells is incorrect. The current implementation marks neighboring cells but does not prevent ships from being placed adjacent to each other. You need to ensure that ships are not placed in neighboring cells by raising an error if a neighboring cell is already occupied.